pkg/controller/apicast/apicast_logic_reconciler.go

```diff
@@ -403,144 +128,110 @@ func (r *APIcastLogicReconciler) applyInitialization() bool {
+		existing.Spec.Template.Spec.ServiceAccountName = desired.Spec.Template.Spec.ServiceAccountName
 	}
 
-	updatedTmp := ReconcileEnvVar(&existingDeployment.Spec.Template.Spec.Containers[0].Env, desiredDeployment.Spec.Template.Spec.Containers[0].Env)
+	updatedTmp := reconcilers.ReconcileEnvVar(&existing.Spec.Template.Spec.Containers[0].Env, desired.Spec.Template.Spec.Containers[0].Env)
 	changed = changed || updatedTmp
 
 	// They are annotations of the PodTemplate, part of the Spec, not part of the meta info of the Pod or Environment object itself
 	// It is not expected any controller to update them, so we use "set" approach, instead of merge.
 	// This way any removed annotation from desired (due to change in CR) will be removed in existing too.
-	if !reflect.DeepEqual(existingDeployment.Spec.Template.Annotations, desiredDeployment.Spec.Template.Annotations) {
+	if !reflect.DeepEqual(existing.Spec.Template.Annotations, desired.Spec.Template.Annotations) {
 		changed = true
-		existingDeployment.Spec.Template.Annotations = desiredDeployment.Spec.Template.Annotations
+		existing.Spec.Template.Annotations = desired.Spec.Template.Annotations
 	}
 
-	if !reflect.DeepEqual(existingDeployment.Spec.Template.Spec.Volumes, desiredDeployment.Spec.Template.Spec.Volumes) {
+	if !reflect.DeepEqual(existing.Spec.Template.Spec.Volumes, desired.Spec.Template.Spec.Volumes) {
 		changed = true
-		existingDeployment.Spec.Template.Spec.Volumes = desiredDeployment.Spec.Template.Spec.Volumes
+		existing.Spec.Template.Spec.Volumes = desired.Spec.Template.Spec.Volumes
 	}
 
-	if !reflect.DeepEqual(existingDeployment.Spec.Template.Spec.Containers[0].VolumeMounts, desiredDeployment.Spec.Template.Spec.Containers[0].VolumeMounts) {
+	if !reflect.DeepEqual(existing.Spec.Template.Spec.Containers[0].VolumeMounts, desired.Spec.Template.Spec.Containers[0].VolumeMounts) {
 		changed = true
-		existingDeployment.Spec.Template.Spec.Containers[0].VolumeMounts = desiredDeployment.Spec.Template.Spec.Containers[0].VolumeMounts
+		existing.Spec.Template.Spec.Containers[0].VolumeMounts = desired.Spec.Template.Spec.Containers[0].VolumeMounts
 	}
 
-	if changed {
-		r.Logger().Info(fmt.Sprintf("Updating %s", k8sutils.ObjectInfo(&existingDeployment)))
-		err = r.Client().Update(context.TODO(), &existingDeployment)
-		return err
-	}
-
-	return nil
+	return changed, nil
 }
 
-func (r *APIcastLogicReconciler) reconcileService(desiredService v1.Service) error {
-	existingService := v1.Service{}
-	err := r.Client().Get(context.TODO(), r.namespacedName(&desiredService), &existingService)
-	if err != nil {
-		if errors.IsNotFound(err) {
-			r.Logger().Info(fmt.Sprintf("Creating %s", k8sutils.ObjectInfo(&desiredService)))
-			err = r.Client().Create(context.TODO(), &desiredService)
-		}
-		return err
+func IngressMutator(existingObj, desiredObj k8sutils.KubernetesObject) (bool, error) {
+	existing, ok := existingObj.(*extensions.Ingress)
```

Comment:
I'm curious about the effects of the type assertion/conversion

I thought type convertion/assertion performs a copy of the object that it tries to convert.

If you make changes, wouldn't they be made to the copy (`existing`) and thus not be effective in `existingObj` when going out of the method?

Reply:
If you run this snippet:

```
import (                                                    
    "encoding/json"                                         
    "fmt"                                                   
)                                                           
                                                            
type I interface {                                          
    Walk()                                                  
}                                                           
                                                            
type S struct {                                             
    B int                                                   
}                                                           
                                                            
func (s *S) Walk() {                                        
}                                                           
                                                            
func updateI(i I) {                                         
    newS, ok := i.(*S)                                      
    if !ok {                                                
        panic("not ok")                                     
    }                                                       
    newS.B = 6                                              
}                                                           
                                                            
func print(s *S) {                                          
    se, _ := json.MarshalIndent(s, "", "\t")                
    fmt.Println(string(se))                                 
}                                                           
                                                            
func main() {                                               
    s := &S{B: 5}                                           
                                                            
    fmt.Println("before")                                   
    print(s)                                                
                                                            
    updateI(s)                                              
                                                            
    fmt.Println("after")                                    
    print(s)                                                
}                                                           

```

The output is:

```
before
{
	"B": 5
}
after
{
	"B": 6
}
```

This is easy to explain. Asserted type is pointer type.

Reply:
Ok, so I understand then that the method expects an interface, and the interface contains a value that's a pointer to an object.

This means that what is copied when doing the assertion is the pointer value (and not what the pointer points to).

This implies that for example you could not replace the entire object itself and get the new object outside of the method because the pointer is copied but you can change the fields of the object :+1: